packages/@aws-cdk/aws-codedeploy/lib/lambda/custom-deployment-config.ts

```diff
@@ -0,0 +1,119 @@
+import { Construct, Fn, Stack } from '@aws-cdk/core';
+import { AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId } from '@aws-cdk/custom-resources';
+import { ILambdaDeploymentConfig } from './deployment-config';
+
+/**
+ * Lambda Deployment config type
+ */
+export enum CustomLambdaDeploymentConfigType {
+  /**
+   * Canary deployment type
+   */
+  CANARY = 'Canary',
+
+  /**
+   * Linear deployment type
+   */
+  LINEAR = 'Linear'
+}
+
+/**
+ * Properties of a reference to a CodeDeploy Lambda Deployment Configuration.
+ */
+export interface CustomLambdaDeploymentConfigProps {
+
+  /**
+   * The type of deployment config, either CANARY or LINEAR
+   */
+  readonly type: CustomLambdaDeploymentConfigType;
+
+  /**
+   * The integer percentage of traffic to shift:
+   * - For LINEAR, the percentage to shift every interval
+   * - For CANARY, the percentage to shift until the interval passes, before the full deployment
+   */
+  readonly percentage: string;
+
+  /**
+   * The interval, in number of minutes:
+   * - For LINEAR, how frequently additional traffic is shifted
+   * - For CANARY, how long to shift traffic before the full deployment
+   */
+  readonly interval: string;
+}
+
+/**
+ * A custom Deployment Configuration for a Lambda Deployment Group.
+ *
+ */
+export class CustomLambdaDeploymentConfig extends Construct implements ILambdaDeploymentConfig {
+
+  /**
+   * The name of the deployment config
+   */
+  public readonly deploymentConfigName: string;
+
+  /**
+   * The arn of the deployment config
+   */
+  public readonly deploymentConfigArn: string;
+
+  public constructor(scope: Construct, id: string, props: CustomLambdaDeploymentConfigProps) {
+    super(scope, id);
+    const deploymentType = 'TimeBased' + props.type.toString();
+    let routingConfig;
+    let customDeploymentConfigName;
+    if (props.type == CustomLambdaDeploymentConfigType.CANARY) {
+      routingConfig = {
+        type: deploymentType,
+        timeBasedCanary: {
+          canaryInterval: props.interval,
+          canaryPercentage: props.percentage,
+        },
+      };
+      customDeploymentConfigName = 'LambdaCanary' + props.percentage.toString() + 'Percent' + props.interval.toString() + 'Minutes';
+    } else if (props.type == CustomLambdaDeploymentConfigType.LINEAR) {
+      routingConfig = {
+        type: deploymentType,
+        timeBasedLinear: {
+          linearInterval: props.interval,
+          linearPercentage: props.percentage,
+        },
+      };
+      customDeploymentConfigName = 'LambdaLinear' + props.percentage.toString() + 'Percent' + props.interval.toString() + 'Minutes';
+    } else {
+      throw new Error('Deployment config type not recognized, only CANARY and LINEAR are supported');
+    }
+
+    // What the config is called in the AWS console
+    const resourceName = id + customDeploymentConfigName;
```

Comment:
How about:

```suggestion
    const resourceName = this.node.uniqueId;
```

instead?

Reply:
This is the name that I use to call the resource in the AWS console. I want it to be called that for a couple of reasons:

1) Readability -- easy to understand exactly what it is supposed to do, and it follows the naming convention of the default configs.
2) Ease of use -- by having the name be a composition of the construct name, deployment type, percentage and interval, if any of those are changed then a new deployment config will be created with the proper settings.
2.1) I included the construct name so a user can have two configs with identical type/percent/interval settings, but as different constructs. This way, if a replacement is needed on one of them, the config for the other construct won't be inadvertently deleted.

Reply:
But the ID is only locally unique, not globally unique?

So if I have:

```ts
new MyConstruct(this, 'Id1');
new MyConstruct(this, 'Id2');

class MyConstruct extends Construct {
  constructor(scope: Construct, id: string) {
    new codedeploy.CustomLambdaDeploymentConfig(this, 'Id', { ... });
  }
}
```

Suddenly you'll create 2 custom deployment configs with the exact same name, no...?

Reply:
I think you're right -- I'll change to use the globally unique id